rasa/utils/train_utils.py

```diff
@@ -0,0 +1,925 @@
+) -> "tf.Tensor":
+    """Use loss depending on given option."""
+
+    if loss_type == "margin":
+        return tf_loss_margin(
+            sim_pos,
+            sim_neg,
+            sim_neg_bot_bot,
+            sim_neg_dial_dial,
+            sim_neg_bot_dial,
+            mask,
+            mu_pos,
+            mu_neg,
+            use_max_sim_neg,
+            C_emb,
+        )
+    elif loss_type == "softmax":
+        return tf_loss_softmax(
+            sim_pos,
+            sim_neg,
+            sim_neg_bot_bot,
+            sim_neg_dial_dial,
+            sim_neg_bot_dial,
+            mask,
+            scale_loss,
+        )
+    else:
+        raise ValueError(
+            "Wrong loss type '{}', "
+            "should be 'margin' or 'softmax'"
+            "".format(loss_type)
+        )
+
+
+# noinspection PyPep8Naming
+def calculate_loss_acc(
+    a_embed: "tf.Tensor",
+    b_embed: "tf.Tensor",
+    b_raw: "tf.Tensor",
+    all_b_embed: "tf.Tensor",
+    all_b_raw: "tf.Tensor",
+    num_neg: int,
+    mask: Optional["tf.Tensor"],
+    loss_type: Text,
+    mu_pos: float,
+    mu_neg: float,
+    use_max_sim_neg: bool,
+    C_emb: float,
+    scale_loss: bool,
+) -> Tuple["tf.Tensor", "tf.Tensor"]:
+    """Calculate loss and accuracy."""
+
+    (
+        pos_dial_embed,
+        pos_bot_embed,
+        neg_dial_embed,
+        neg_bot_embed,
+        dial_bad_negs,
+        bot_bad_negs,
+    ) = sample_negatives(a_embed, b_embed, b_raw, all_b_embed, all_b_raw, num_neg)
+
+    # calculate similarities
+    (sim_pos, sim_neg, sim_neg_bot_bot, sim_neg_dial_dial, sim_neg_bot_dial) = tf_sim(
+        pos_dial_embed,
+        pos_bot_embed,
+        neg_dial_embed,
+        neg_bot_embed,
+        dial_bad_negs,
+        bot_bad_negs,
+        mask,
+    )
+
+    acc = tf_calc_accuracy(sim_pos, sim_neg)
+
+    loss = choose_loss(
+        sim_pos,
+        sim_neg,
+        sim_neg_bot_bot,
+        sim_neg_dial_dial,
+        sim_neg_bot_dial,
+        mask,
+        loss_type,
+        mu_pos,
+        mu_neg,
+        use_max_sim_neg,
+        C_emb,
+        scale_loss,
+    )
+
+    return loss, acc
+
+
+def confidence_from_sim(sim: "tf.Tensor", similarity_type: Text) -> "tf.Tensor":
+    if similarity_type == "cosine":
+        # clip negative values to zero
+        return tf.nn.relu(sim)
+    else:
+        # normalize result to [0, 1] with softmax
+        return tf.nn.softmax(sim)
+
+
+def linearly_increasing_batch_size(
+    epoch: int, batch_size: Union[List[int], int], epochs: int
+) -> int:
+    """Linearly increase batch size with every epoch.
+
+    The idea comes from https://arxiv.org/abs/1711.00489.
+    """
+
+    if not isinstance(batch_size, list):
+        return int(batch_size)
+
+    if epochs > 1:
+        return int(
+            batch_size[0] + epoch * (batch_size[1] - batch_size[0]) / (epochs - 1)
+        )
+    else:
+        return int(batch_size[0])
+
+
+def output_training_stat_dataset(
```

Comment:
Can we rename this to `output_validation_stat`, since this is now true metrics on validation datasets?

Reply:
good catch, thanks